File: leetcodes/211. Add and Search Word - Data structure design/WordDictionary.java
class TrieNode { //slower than js solution
  //links array with TrieNode type
  public TrieNode[] links = new TrieNode[26];

  //boolean isEnd for the end of words
    public boolean isEnd;
  //initialize constructor
    public TrieNode() {}
}

class WordDictionary {
  private TrieNode root;
  //constructor
  public WordDictionary() {
    root = new TrieNode();
  }
  public void addWord(String word){
    TrieNode curNode = root;
    for (int i = 0; i<word.length(); i++) {
      char c = word.charAt(i);
      if (curNode.links[c-'a'] ==  null) {
        curNode.links[c-'a'] = new TrieNode();
      }
      curNode = curNode.links[c-'a'];
    }
    curNode.isEnd = true;
  }
  public boolean search(String word){
    return match(word, 0, root);
  }

  //back track
  private boolean match(String word, int index, TrieNode curNode){
    //if index equals to word counts, it means over the boundary -> return current node's isEnd value
    if (index == word.length()) {
      return curNode.isEnd;
    }
    //if current chartAtindex is '.'
    if (word.charAt(index) != '.') {
      if (curNode.links[word.charAt(index) - 'a'] != null) {
        //if current element is not null
        return match(word, index+1, curNode.links[word.charAt(index) - 'a']);
        //return search next index of the word with current node
      }
      //else
    } else {
      //boolean result = true;
      //iterate over the current node links[26]
      for (int i = 0; i < curNode.links.length; i++) {
        //if current element is not null
        if (curNode.links[i] != null && match(word, index + 1, curNode.links[i])) {
          return true;
          //check with all occurrences of the not null character when . is encountered then return search next index of word, with current element
        }
      }
    }
    return false;
  }

}

File: leetcodes/211. Add and Search Word - Data structure design/wordDictionary.ts
class TrieNode {
  isEnd: boolean
  children: Array<TrieNode>
  constructor() {
    this.children= new Array(26)
    this.isEnd = false
  }
}


class WordDictionary {
  root: TrieNode
  constructor() {
    this.root = new TrieNode()
  }

  addWord(word: string): void {
    let node = this.root
    for (let i = 0; i <= word.length-1 ; i++) {
      let c: number = word.charCodeAt(i)
      let index: number = c - 'a'.charCodeAt(0)
      if (node.children[index] == null) {
        node.children[index] = new TrieNode()
      }
      node = node.children[index]
    }
      node.isEnd = true
  }

  search(word: string): boolean {
    let node = this.root
    return this.match(word, 0, node)
  }
  match(word: string, idx:number, curNode: TrieNode):boolean{
    //check isEnd when finished word iteration
    if (word.length == idx) return curNode.isEnd
    //if not .
    if (word[idx] != '.') {
      //check if curNode contains word[idx]
      let c: number = word.charCodeAt(idx)
      let index: number = c - 'a'.charCodeAt(0)
      if(curNode.children[index]) {
        return this.match(word, idx+1, curNode.children[index])
      }
      //return match(word, idx+1, curNode.children[index])
      //else false
      //else
    } else {
      //iterate over curNode.children
      for (let i = 0; i < curNode.children.length; i++ ){
        //if curNode.children[i] && match(word, idx+1, curNode.children[i])
        if (curNode.children[i] && this.match(word, idx+1, curNode.children[i]))
        return true
      }
       //return true
       //else false
    }
    return false
  }
}

/**
* Your WordDictionary object will be instantiated and called as such:
* var obj = new WordDictionary()
* obj.addWord(word)
* var param_2 = obj.search(word)
*/

//above runtime error for out of memory? using array and recursive way

class WordDictionary {
  isWord: boolean;
  child: {[Key: string]: WordDictionary};
  constructor() {
      this.child = {};
      this.isWord = false;
  }

  addWord(word: string): void {
      let curr: WordDictionary = this;
      for(const c of word) {
          if(!curr.child[c]) {
              curr.child[c] = new WordDictionary();
          }
          curr = curr.child[c];
      }
      curr.isWord = true;
  }

  search(word: string, i = 0): boolean {
      let curr: WordDictionary = this;
      for(;i < word.length; i++) {
          const c = word[i];
          if(c !== '.') {
              if(!curr.child[c]) {
                  return false;
              }
              curr = curr.child[c];
          } else {
              for(const key in curr.child) {
                  if(curr.child[key].search(word, i + 1)){
                      return true;
                  };
              }
              return false;
          }
      }
      return curr.isWord;
  }
}